rejex.mjs

```diff
@@ -0,0 +1,172 @@
+    // tokens before we pass the remainder into the right.
+    const [lhs_is_match, lhs_remainder] = this.lhs.match(tokens)
+    const [rhs_is_match, rhs_remainder] = this.rhs.match(lhs_remainder)
+
+    // If either subexpressions didn't match, this expression does not match.
+    if (!lhs_is_match || !rhs_is_match) return [false, tokens]
+
+    // This expression is a match. Return the tokens remaining after both the
+    // left and right subexpressions consume them.
+    return [true, rhs_remainder]
+  }
+}
+
+// Alternate two subexpressions together. Matches if either of the
+// subexpressions match.
+export class AlternationExpr {
+  constructor(lhs, rhs) {
+    this.lhs = lhs
+    this.rhs = rhs
+  }
+
+  // Match against a list of character tokens.
+  //
+  // The return value is always a two-element list consisting of
+  // [is_match, remaining_tokens]. The first element is always a bool, and the
+  // second is always a list of tokens.
+  // If there is a match (is_match is true), the matching tokens will be
+  // removed from the remaining_tokens list. Otherwise, the original tokens is
+  // returned in remaining_tokens.
+  match(tokens) {
+    // Try matching both the left and right subexpressions.
+    //
+    // Since there's no ordering to alternation (e.g. [ab] is equivalent to
+    // [ba]), we don't pass the consumed tokens from the one side into the
+    // other.
+    const [lhs_is_match, lhs_remainder] = this.lhs.match(tokens)
+    const [rhs_is_match, rhs_remainder] = this.rhs.match(tokens)
+
+    // If either the left or the right matched, we match
+    if (lhs_is_match || rhs_is_match) {
+      // We can return either the lhs_remainder or the rhs_remainder. We
+      // choose to return the shorter of the two remainders. By doing so,
+      // this expression "consumes" as much of the input as possible, and
+      // therefore is considered "greedy".
+      let shorter_remainder = lhs_remainder
+      if (lhs_remainder.length > rhs_remainder.length) {
+        shorter_remainder = rhs_remainder
+      }
+      return [true, shorter_remainder]
+    }
+
+    // There's no match.
+    return [false, tokens]
+  }
+}
+
+// Repeat a subexpression. Matches if the subexpression matches *zero or more*
+// times.
+export class RepetitionExpr {
+  constructor(expr) {
+    this.expr = expr
+  }
+
+  // Match against a list of character tokens.
+  //
+  // The return value is always a two-element list consisting of
+  // [is_match, remaining_tokens]. The first element is always a bool, and the
+  // second is always a list of tokens.
+  // If there is a match (is_match is true), the matching tokens will be
+  // removed from the remaining_tokens list. Otherwise, the original tokens is
+  // returned in remaining_tokens.
+  match(tokens) {
+    let remainder = tokens
+    let is_match = true
+    // Repetitively apply the subexpression until it doesn't match anymore.
```

Comment:
Maybe say "consume" instead of "apply". Again, trying to make things more obvious. Some people will hesitate for a bit on code like this first thinking, wait, is it actually doing something? It's returning true for is_match all the time. Then after a moment they realize ohhh, the while loop is actually shortening/consuming the remainder... Or at least just make what I mentioned clear in the comments.

Reply:
Consume here isn't quite right here I think... you *apply* an expression, and that expression *consumes* its input.

Reply:
Oh right, you are totally correct in terms of the comment English itself. I think I was considering the broader context of what's happening here and to place some emphasis on it. Like, "// Repetitively apply the subexpression until it doesn't match anymore. The while loop will keep eating/consuming/shortening the remaining expressions and characters until is_match is false." Something like that.